lighthouse-core/gather/computed/first-interactive.js

```diff
@@ -0,0 +1,203 @@
+  get name() {
+    return 'FirstInteractive';
+  }
+
+  /**
+   * @param {number} t The time passed since FMP in miliseconds.
+   * @return {number}
+   */
+  static getRequiredWindowSizeInMs(t) {
+    const tInSeconds = t / 1000;
+    const exponentiationComponent = Math.exp(-.045 * tInSeconds);
+    return (4 * exponentiationComponent + 1) * 1000;
+  }
+
+  /**
+   * @param {!Array<{start: number, end: number}>} tasks
+   * @param {number} startIndex
+   * @param {number} windowEnd
+   * @return {!Array<{start: number, end: number, duration: number}>}
+   */
+  static getTaskGroupsInWindow(tasks, startIndex, windowEnd) {
+    const groups = [];
+
+    let lastEndTime = -Infinity;
+    let currentGroup = null;
+
+    // consider all tasks that could possibly be part of a group starting before windowEnd
+    const considerationWindowEnd = windowEnd + MIN_TASK_GROUP_PADDING;
+    const couldTaskBeInGroup = task => task && task.start < considerationWindowEnd;
+    for (let i = startIndex; couldTaskBeInGroup(tasks[i]); i++) {
+      const task = tasks[i];
+
+      // check if enough time has passed to start a new group
+      if (task.start - lastEndTime > MIN_TASK_GROUP_PADDING) {
+        if (currentGroup) {
+          groups.push(currentGroup);
+        }
+
+        currentGroup = [];
+      }
+
+      currentGroup.push(task);
+      lastEndTime = task.end;
+    }
+
+    if (currentGroup) {
+      groups.push(currentGroup);
+    }
+
+    return groups
+      // add some useful information about the group
+      .map(tasks => {
+        const start = tasks[0].start;
+        const end = tasks[tasks.length - 1].end;
+        const duration = end - start;
+        return {start, end, duration, tasks};
+      })
+      // filter out groups that started after the window
+      .filter(group => group.start < windowEnd);
+  }
+
+  /**
+   * @param {number} FMP
+   * @param {number} traceEnd
+   * @param {!Array<{start: number, end: number>}} longTasks
+   * @return {number}
+   */
+  static findQuietWindow(FMP, traceEnd, longTasks) {
+    if (longTasks.length === 0 ||
+        longTasks[0].start > FMP + MAX_QUIET_WINDOW_SIZE) {
+      return FMP;
+    }
+
+    // FirstInteractive must start at the end of a long task, consider each long task and
+    // examine the window that follows it.
+    for (let i = 0; i < longTasks.length; i++) {
+      const event = longTasks[i];
+      const windowStart = event.end;
+      const windowSize = FirstInteractive.getRequiredWindowSizeInMs(windowStart - FMP);
+      const windowEnd = windowStart + windowSize;
+
+      // Check that we have a long enough trace
+      if (windowEnd > traceEnd) {
+        throw new Error(TRACE_BUSY_MSG);
+      }
+
+      const isTooCloseToFMP = group => group.start < FMP + MIN_TASK_GROUP_FMP_DISTANCE;
+      const isTooLong = group => group.duration > MAX_TASK_GROUP_DURATION;
+      const isBadTaskGroup = group => isTooCloseToFMP(group) || isTooLong(group);
```

Comment:
I agree it's a bit much and I started with inlining when my names were shorter, but inlining now creates a linebreak and seems marginally better to keep the two conditions on the same line and keep each item a one-liner. still prefer a two-line inliner over this?